### PR TITLE
feat: expose settle deal command for lotus miner

### PR DESCRIPTION
**Pull request by akaladarshi**
## Related Issues
Fixes: #11590 

## Proposed Changes
Expose settle-deal command for lotus-miner

### CODE REVIEW

**Comment by rjan90** [comment]
@akaladarshi Thanks for working on this. Two initial things based on the failing checks:

- In you /lotus-folder, you need to run `make gen` and `make docsgen-cli` for the gen-check and docs-check to pass.
- Add a entry in the unreleased section of the [Changelog](https://github.com/filecoin-project/lotus/blob/master/CHANGELOG.md) for what this feature does.

**Comment by akaladarshi** [comment]
> @akaladarshi Thanks for working on this. Two initial things based on the failing checks:
> 
> * In you /lotus-folder, you need to run `make gen` and `make docsgen-cli` for the gen-check and docs-check to pass.
> * Add a entry in the unreleased section of the [Changelog](https://github.com/filecoin-project/lotus/blob/master/CHANGELOG.md) for what this feature does.

Hey @rjan90, Thanks for informing.

I just wanted to get an idea if this is what was expected for the implementation, so raised a draft PR.
Once I get confirmation I will clean the PR and fix all the failing CI checks.

**Review by LexLuthr — CHANGES_REQUESTED** [review]
We seem to be missing the case for settling all deal payments.
https://github.com/filecoin-project/lotus/issues/11590#issuecomment-2057438506

**Comment by rjan90** [comment]
@LexLuthr Want to give an updated review of this PR after the code changes for settling all deal payments landed.

File: cli/spcli/actor.go [hunk 1]
```diff
@@ -34,6 +36,112 @@ import (
...
+					return err
+				}
+
+				// if all the deals were successful
+				if settlementReturn.Results.SuccessCount == uint64(len(dealIDs)) {
+					fmt.Printf("Successfully settled %d deals\n", settlementReturn.Results.SuccessCount)
+				} else {
+					fmt.Printf("Settled %d out of %d deals\n", settlementReturn.Results.FailCount, len(dealIDs))
+				}
+			}
```
Comment (LexLuthr) [inline cli/spcli/actor.go:139]: We don't need the if statement as method doesn't exists before NV23. We should directly try to unmarshal the return. 
FYI: go-state-type version != network.Version

File: cli/spcli/actor.go [hunk 2]
```diff
@@ -34,6 +36,112 @@ import (
...
+				var settlementReturn markettypes14.SettleDealPaymentsReturn
+				if err = settlementReturn.UnmarshalCBOR(bytes.NewReader(wait.Receipt.Return)); err != nil {
+					return err
+				}
+
+				// if all the deals were successful
+				if settlementReturn.Results.SuccessCount == uint64(len(dealIDs)) {
+					fmt.Printf("Successfully settled %d deals\n", settlementReturn.Results.SuccessCount)
+				} else {
+					fmt.Printf("Settled %d out of %d deals\n", settlementReturn.Results.FailCount, len(dealIDs))
```
Comment (LexLuthr) [inline cli/spcli/actor.go:137]: Incorrect success count calculation.

```suggestion
					fmt.Printf("Settled %d out of %d deals\n", len(dealIDs) - settlementReturn.Results.FailCount, len(dealIDs))
```

File: cli/spcli/actor.go [hunk 3]
```diff
@@ -34,6 +36,112 @@ import (
...
+				}
+			}
+			return nil
```
Comment (LexLuthr) [inline cli/spcli/actor.go:140]: We should display the settlement amount as well. User needs to know if they spent more gas than the actual settlement amount.

Comment (LexLuthr) [reply cli/spcli/actor.go:140]: We should probably also show how many deals have completed their lifetime

**Review by LexLuthr — CHANGES_REQUESTED** [review]
Sorry for the delayed review.
Logic is sound. Just need some changes to the result output statements.

File: cli/spcli/actor.go (hunk 2 above)
Comment (akaladarshi) [reply cli/spcli/actor.go]: Done.

Just a question is there any other state the deal could be in apart from `success` and `fail`?
If not then we can do something like this: 
`fmt.Printf("Settled %d out of %d deals\n", settlementReturn.Results.SuccessCount, len(dealIDs))`

File: cli/spcli/actor.go (hunk 3 above)
Comment (akaladarshi) [reply cli/spcli/actor.go:148]: Done.

File: cli/spcli/actor.go (hunk 2 above)
Comment (LexLuthr) [reply cli/spcli/actor.go]: Return type is already predefined so it will always stick to those types. Your suggestion seems better than current one.

Comment (LexLuthr) [reply cli/spcli/actor.go]: I will approve once this final change is finished.

Comment (akaladarshi) [reply cli/spcli/actor.go]: Done.

**Review by LexLuthr — APPROVED** [review]

**Comment by beck-8** [comment]
How can users distinguish whether they have already executed manual settlement? Is there no way to distinguish? Does that mean that if the user has 100,000 transactions, it must be executed atomically, otherwise he doesn't know which part he executes? Will repeated execution waste gas?

**Pull request merged**